File: flut/.flutter/lib/flutter/widgets/interactive_viewer.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flutter/foundation/change_notifier.dart';

class FlutPanAxis extends FlutEnumObject<PanAxis> {
  const FlutPanAxis()
    : super('PanAxis', const {
        'horizontal': PanAxis.horizontal,
        'vertical': PanAxis.vertical,
        'aligned': PanAxis.aligned,
        'free': PanAxis.free,
      });
}

class FlutInteractiveViewer {
  FlutInteractiveViewer._();

  static InteractiveViewer? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return InteractiveViewer(
      key: runtime.decodeKey(data),
      clipBehavior: runtime.unpackRequiredField<Clip>(data, 'clipBehavior'),
      panAxis: runtime.unpackRequiredField<PanAxis>(data, 'panAxis'),
      boundaryMargin: runtime.unpackRequiredField<EdgeInsets>(
        data,
        'boundaryMargin',
      ),
      constrained: runtime.unpackRequiredField<bool>(data, 'constrained'),
      maxScale: runtime.unpackRequiredField<double>(data, 'maxScale'),
      minScale: runtime.unpackRequiredField<double>(data, 'minScale'),
      interactionEndFrictionCoefficient: runtime.unpackRequiredField<double>(
        data,
        'interactionEndFrictionCoefficient',
      ),
      onInteractionEnd: runtime.unpackOptionalCallback(
        data,
        'onInteractionEnd',
      ),
      onInteractionStart: runtime.unpackOptionalCallback(
        data,
        'onInteractionStart',
      ),
      onInteractionUpdate: runtime.unpackOptionalCallback(
        data,
        'onInteractionUpdate',
      ),
      panEnabled: runtime.unpackRequiredField<bool>(data, 'panEnabled'),
      scaleEnabled: runtime.unpackRequiredField<bool>(data, 'scaleEnabled'),
      scaleFactor: runtime.unpackRequiredField<double>(data, 'scaleFactor'),
      transformationController: runtime
          .unpackOptionalField<TransformationController>(
            data,
            'transformationController',
          ),
      trackpadScrollCausesScale: runtime.unpackRequiredField<bool>(
        data,
        'trackpadScrollCausesScale',
      ),
      alignment: runtime.unpackOptionalField<Alignment>(data, 'alignment'),
      child: runtime.unpackRequiredField<Widget>(data, 'child'),
    );
  }
}

class FlutTransformationController
    extends FlutChangeNotifier<TransformationController> {
  FlutTransformationController.createFromData({
    required super.runtime,
    required super.data,
    required super.target,
  }) : super.createFromData();

  FlutTransformationController.createFromObject({
    required super.runtime,
    required super.oid,
    required super.target,
  }) : super.createFromObject(type: 'TransformationController');

  static FlutRealtimeObject flutCreate(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    final value = runtime.unpackOptionalField<Matrix4>(data, 'value');
    return FlutTransformationController.createFromData(
      runtime: runtime,
      data: data,
      target: TransformationController(value ?? Matrix4.identity()),
    );
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'value':
        return flutTarget.value;
    }
    return super.getRawProperty(property);
  }

  @override
  bool setProperty(String property, dynamic value) {
    switch (property) {
      case 'value':
        flutTarget.value = value as Matrix4;
        return true;
    }
    return super.setProperty(property, value);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'toScene':
        return flutTarget.toScene(args[0] as Offset);
    }
    return super.callMethod(method, args, kwargs);
  }
}

File: flut/flutter/widgets/interactive_viewer.py
from typing import Callable, Optional, override

from flut._flut.engine import FlutEnumObject, _flut_pack_value
from flut.dart.ui import Clip
from flut.flutter.foundation.change_notifier import Listenable, ValueNotifier
from flut.flutter.painting.edge_insets import EdgeInsets
from flut.flutter.widgets.framework import Widget


class PanAxis(FlutEnumObject):
    horizontal: "PanAxis"
    vertical: "PanAxis"
    aligned: "PanAxis"
    free: "PanAxis"


class InteractiveViewer(Widget):
    _flut_type = "InteractiveViewer"

    def __init__(
        self,
        *,
        key=None,
        clipBehavior=Clip.hardEdge,
        panAxis: PanAxis = PanAxis.free,
        boundaryMargin=EdgeInsets.zero,
        constrained: bool = True,
        maxScale: float = 2.5,
        minScale: float = 0.8,
        interactionEndFrictionCoefficient: float = 0.0000135,
        onInteractionEnd: Optional[Callable] = None,
        onInteractionStart: Optional[Callable] = None,
        onInteractionUpdate: Optional[Callable] = None,
        panEnabled: bool = True,
        scaleEnabled: bool = True,
        scaleFactor: float = 200.0,
        transformationController: Optional["TransformationController"] = None,
        trackpadScrollCausesScale: bool = False,
        alignment=None,
        child: Optional[Widget] = None,
    ):
        super().__init__(key=key)
        self.clipBehavior = clipBehavior
        self.panAxis = panAxis
        self.boundaryMargin = boundaryMargin
        self.constrained = constrained
        self.maxScale = maxScale
        self.minScale = minScale
        self.interactionEndFrictionCoefficient = interactionEndFrictionCoefficient
        self.onInteractionEnd = onInteractionEnd
        self.onInteractionStart = onInteractionStart
        self.onInteractionUpdate = onInteractionUpdate
        self.panEnabled = panEnabled
        self.scaleEnabled = scaleEnabled
        self.scaleFactor = scaleFactor
        self.transformationController = transformationController
        self.trackpadScrollCausesScale = trackpadScrollCausesScale
        self.alignment = alignment
        self.child = child

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["panAxis"] = _flut_pack_value(self.panAxis)
        result["constrained"] = _flut_pack_value(self.constrained)
        result["maxScale"] = _flut_pack_value(self.maxScale)
        result["minScale"] = _flut_pack_value(self.minScale)
        result["interactionEndFrictionCoefficient"] = _flut_pack_value(
            self.interactionEndFrictionCoefficient
        )
        result["panEnabled"] = _flut_pack_value(self.panEnabled)
        result["scaleEnabled"] = _flut_pack_value(self.scaleEnabled)
        result["scaleFactor"] = _flut_pack_value(self.scaleFactor)
        result["trackpadScrollCausesScale"] = _flut_pack_value(
            self.trackpadScrollCausesScale
        )
        result["clipBehavior"] = _flut_pack_value(self.clipBehavior)
        result["boundaryMargin"] = _flut_pack_value(self.boundaryMargin)
        if self.onInteractionEnd is not None:
            result["onInteractionEnd"] = self._register_action(
                self.onInteractionEnd, "GestureScaleEndCallback"
            )._flut_pack()
        if self.onInteractionStart is not None:
            result["onInteractionStart"] = self._register_action(
                self.onInteractionStart, "GestureScaleStartCallback"
            )._flut_pack()
        if self.onInteractionUpdate is not None:
            result["onInteractionUpdate"] = self._register_action(
                self.onInteractionUpdate, "GestureScaleUpdateCallback"
            )._flut_pack()
        if self.transformationController is not None:
            result["transformationController"] = _flut_pack_value(
                self.transformationController
            )
        if self.alignment is not None:
            result["alignment"] = _flut_pack_value(self.alignment)
        if self.child is not None:
            result["child"] = _flut_pack_value(self.child)
        return result


class TransformationController(ValueNotifier):
    _flut_type = "TransformationController"

    def __init__(self, value=None):
        props = {}
        if value is not None:
            props["value"] = _flut_pack_value(value)
        self._flut_init_props = props
        Listenable.__init__(self)

    def toScene(self, viewportPoint):
        return self._flut_call("toScene", viewportPoint._flut_pack())
